### PR TITLE
registry apply YAML import of deployments, versions, specs, and non-project artifacts

### PR DESCRIPTION
**Pull request by timburks**
Addresses #771. See the description there for desired behavior. Notably, this adds a new `parent` field to `metadata` sections to specifies the parent resource of resources described in YAML. For versions this is `api/{api_id}`. Note that the containing project and location are omitted from `metadata.parent` values. This allows YAML exports to be portable across projects.

In the future, we might want to support other variations in the `parent` field, such as containing the full project-rooted resource name (although I personally think this should be discouraged) or having abbreviated forms like `{api_id}/{version_id}` instead of `apis/{api_id}/versions/{version_id}` (I think this also has problems: ambiguity with deployments and increased inconsistency with the Registry API.)

It could be argued that omitting `project/{project_id}/location/{location_id}` from `metadata.parent` values is also inconsistent with the API, but I think the benefits of portable resource YAMLs justifies that.

This allows bulk application of directories of YAML files and to ensure that all parents are created before their children, resources are all read before application and are applied in a specific order: first APIs, then versions, specs, deployments, and finally artifacts.

Two new tests upload the same project that is described in the `samples` directory but as "unbundled" files that are stored in two different directory organizations: one flat, one hierarchical.

### CODE REVIEW

File: cmd/registry/patch/version.go [hunk 1]
```diff
@@ -66,15 +66,39 @@ func newApiVersion(ctx context.Context, client *gapic.RegistryClient, message *r
 	}, nil
 }
 
+func applyApiVersionPatchBytes(
+	ctx context.Context,
+	client connection.RegistryClient,
+	bytes []byte,
+	parent string) error {
```
Comment (seaneganx) [inline cmd/registry/patch/version.go:73]: This function can easily be defined in one line. We have much longer function definitions throughout the codebase, and defining it on one line prevents indentation confusion. This style causes the function definition to blend in with the code it contains. More details messaged privately.

File: cmd/registry/cmd/apply/apply_test.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ import (
 )
 
 const sampleDir = "testdata/sample"
+const unbundledHierarchicalSampleDir = "testdata/unbundled-hierarchical"
+const unbundledFlatSampleDir = "testdata/unbundled-flat"
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/cmd/apply/apply_test.go]: OK, I made these inline.

File: cmd/registry/cmd/apply/apply_test.go [hunk 3]
```diff
@@ -151,101 +153,123 @@ func TestApply(t *testing.T) {
...
+			expected, err := os.ReadFile(filename)
+			if err != nil {
+				t.Fatalf("Failed to read API YAML: %s", err)
+			}
+
+			got, err := registryClient.GetApi(ctx, &rpc.GetApiRequest{
+				Name: project.Api("registry").String(),
+			})
+			if err != nil {
+				t.Fatalf("failed to get api: %s", err)
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/cmd/apply/apply_test.go]: Fixed

File: cmd/registry/patch/apply.go [hunk 4]
```diff
@@ -26,29 +26,83 @@ import (
...
+func (p *patchGroup) run(ctx context.Context, jobs int) error {
+	// Apply each resource type independently in order of ownership (parents first).
+	for _, tasks := range [][]core.Task{
+		p.apiTasks,
+		p.versionTasks,
+		p.specTasks,
+		p.deploymentTasks,
+		p.artifactTasks,
+	} {
+		if len(tasks) > 0 {
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/patch/apply.go]: OK, this is removed.

File: cmd/registry/patch/version.go (hunk 1 above)
Comment (timburks) [reply cmd/registry/patch/version.go:73]: Thanks for the link. As written this is consistent with at least one other instance in this file and several more in parallel. I think those should be addressed together (and separately). I filed #803 and there counted 46 instances of multiline declarations.

**Review by seaneganx — APPROVED** [review]

**Pull request merged**